---
title: "Running a simulation based on SRT data region"
output: 
  html_document:
    number_sections: true
    toc: true
pkgdown:
  as_is: true
vignette: >
  %\VignetteIndexEntry{Running a simulation based on SRT data region}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This tutorial will run a simulation with `sCCIgen` without the interactive interface based on SRT data. 

# Load R package 

```{r, eval=FALSE}
library(sCCIgen)
```

# Load and clean sample data 

Download sample data from https://github.com/songxiaoyu/sCCIgen_data/tree/main/input_data

```{r, eval=FALSE}
load("SeqFishPlusCortex_2025_expr.Rdata")
load("SeqFishPlusCortex_2025_spatial.Rdata")

dim(expr)
expr[1:3, 1:3]
dim(spatial)

anno <- colnames(expr)
region <- spatial[, 4]
```

# Analysis of the existing data to provide insights into the parameters of the simulation

Users can split the `sCCIgen` simulation into (1) <u> model fitting </u> and (2) <u> simulation using fitted model and user-provided parameters </u> steps to expedite the simulations. 

It is especially helpful if the number of genes and/or cells are very large and users want to run simulation for more than once.  By splitting the simulation into these two steps, users can estimate model parameters only once and save the results for multiple use. 

## Task 1: Estimate model parameters from the snRNAseq for simulation

This is part is to fit the expression data. When sim_method=="copula", it will fit both the gene marginal distribution and gene-gene correlation. When sim_method=="ind", it will only fit the gene marginal distribution. 

Note: If the number of genes or cells are large, model fitting may take some time. It is suggested to select a reasonable sample size (e.g. <10000 per cell type) before the model fitting, as more cells may not be needed improve the estimation. Similarly, if some genes are extremely zero-inflated, narrowing the simulation to reasonably variable genes is an option. 

```{r, eval=FALSE}
# model fitting
ModelEst <- Est_ModelPara(expr = expr, 
                          anno = anno, 
                          sim_method = "ind", 
                          region = region, 
                          ncores = 10)

saveRDS(ModelEst, file = "SeqFishPlusCortex_fit_wo_cor_region.RDS")
```

## Task 2: Estimate CCIs in the input data following the existing Giotto pipeline

```{r, eval=FALSE}
library(Giotto)

# Preprocess data with Giotto pipeline
db <- preprocessGiotto(expr_data = expr, 
                       spatial_data = spatial, 
                       run_hvg = TRUE,
                       run_kNN_network = TRUE,
                       run_Delaunay_network = TRUE)

# Estimate cell-cell attraction and inhibition patterns, and save in 
# pre-defined folder
cellProximityTable(gobject = db, 
                   abs_enrichm = 0.3, 
                   p_adj = 0.05,
                   output_file = "est_CCI_dist_dist_region.csv")

# Estimate gene expressions of cells impacted by their neighbors, and save in
# pre-defined folder
ExprDistanceTable(gobject = db, 
                  in_hvg = TRUE, 
                  region_specific = TRUE, 
                  abs_log2fc_ICG = 0.3, 
                  p_adj = 0.05,
                  output_file = "est_CCI_dist_expr_region.csv")

# Estimate gene expressions of cells impacted by gene expressions of neighboring # cells, narrow to known pairs, such as ligand and receptor pairs, and save in
# pre-defined folder
ExprExprTable(gobject = db, 
              database = "mouse", 
              region_specific = TRUE, 
              p_adj = 0.05, 
              abs_log2fc_LR = 0.3,
              output_file = "est_CCI_expr_expr_region.csv")
```


## Task 3: Estimate spatial region specific genes

```{r, eval=FALSE}
# Estimate spatial region specific genes.
SpatialTable(gobject = db, 
             top_num = 2, 
             fdr_cut = 0.05, 
             output_file = "est_CCI_expr_expr_region.csv")
```

# Create a parameter file

Users need to create a parameter file. The sample parameter file for snRNAseq based simulation is [here](https://github.com/songxiaoyu/sCCIgen_data/tree/main/sample_parameter_file/SRT) for downloading and filling in to perform simulations. 
 
# Perform the entire simulation and save the results

Assuming you already have a parameter file, you can run the entire simulation using codes like this:

```{r, eval=FALSE}
model_param_path <- "SeqFishPlusCortex_fit_wo_cor_region.RDS"

input <- "SeqFishPlus_Region_AllPattern.tsv"

ParaSimulation(input = input, ModelFitFile = model_param_path)
```



